lib-extra/src/main/java/com/diffplug/spotless/extra/config/MavenCoordinates.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2016 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.extra.config;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Serializable;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+/** Configuration of Maven coordinates */
+class MavenCoordinates implements Serializable {
+
+	// Not used, only the serialization output is required to determine whether the object has changed
+	private static final long serialVersionUID = 1L;
+	private final Set<Coordinate> coordinates;
+
+	/** Creates empty coordinate set */
+	MavenCoordinates() {
+		this.coordinates = new TreeSet<Coordinate>(new DependencyComparator());
+	}
+
+	/** Get all dependency restrictions */
+	String[] get() {
```

Comment:
Effective Java says to prefer lists over arrays, unless e.g. the performance gained from using an array is measurably significant and truly needed.

So unless you have numbers from a JMH benchmark to prove that we need to return `String[]` here, I would heartily recommend returning an unmodifiable `List<String>` instead, e.g.:

```java
List<String> get() {
    return coordinates
        .stream()
        .map(c -> c.toString())
        .collect(toCollectionAndThen(ArrayList::new, Collections::unmodifiableList));
}
```

Reply:
I prefer to keep the get/add/update interfaces aligned. Since the add/update work with native array, I would prefer to stick with the approach. Agreed?

Reply:
> The only user for get (beside tests) is the JarState.

Okay, fair point! But I personally prefer returning an unmodifiable collection because IMO it keeps things conceptually simpler. :)

> As part of my changes, I also refactored it to support "varargs". It did not support lists before and I thought it would be a nice way to avoid dedicated methods for a single object and a collection.

I admit I'm not such a fan that the varargs forced the "mavenCoordinates" parameter to be moved to the end of the parameter list of `JarState.from()` in this case.

So to reduce potential confusion for new people reading over JarState's constructors and `JarState.from()`, I would revert this refactoring and introduce a new `JarState.from(Iterable<String> mavenCoordinates, Provisioner provisioner)` static factory method instead. :)

Reply:
It would also eliminate the need to `@Deprecate` `JarState.from(String, Provisioner)`, AFAICT.